### PR TITLE
Improving broadcasting performance by working around recursion limits of inlining

### PR DESCRIPTION
**Pull request by mateuszbaran**
This PR implements the changes proposed in #41090 . All it does is duplicating the two methods that are marked as deletions in this PR. Benchmarks that motivate this change are available here: https://github.com/JuliaLang/julia/issues/41090#issuecomment-855804337 .

### CODE REVIEW

**Comment by mbauman** [comment]
I must confess it's _always_ a challenge for me to understand the existing recursive magic whenever I look at this file.  I can often get there, but it makes my head hurt a bit.

I understand the motivation here, but the indirection makes this intractable for me.  🤯

**Comment by mateuszbaran** [comment]
I still don't understand how `make_makeargs` works, I just noticed that these specific methods weren't inlined sometimes and found a way to fix that :slightly_smiling_face: .

**Comment by mateuszbaran** [comment]
Bump?

**Comment by Pramodh-G** [comment]
Bump

**Comment by oscardssmith** [comment]
I think we need triage to decide whether the performance gain is worth the added complexity.

File: base/broadcast.jl [hunk 1]
```diff
@@ -342,6 +342,47 @@ cat_nested(t::Broadcasted, rest...) = (cat_nested(t.args...)..., cat_nested(rest
...
+struct RecursiveInliningEnforcerB{TMT,TMH,TT,TH,TF}
+    makeargs_tail::TMT
+    makeargs_head::TMH
+    headargs::TH
+    tailargs::TT
+    f::TF
+end
+
+for UB in [Any, RecursiveInliningEnforcerA]
+    @eval @inline function (bb::RecursiveInliningEnforcerB{TMT,TMH,TT,TH,TF})(args::Vararg{Any,N}) where {N,TMT,TMH<:$UB,TT,TH,TF}
```
Comment (JeffBezanson) [inline base/broadcast.jl:372]: Why does this need two methods? Doesn't the `Any` bound cover everything?

**Comment by JeffBezanson** [comment]
Triage thinks this is a bit too dependent on current details of how the compiler works. It basically just names the types of the returned functions and adds unnecessary extra methods. From looking at the code, there is no good reason why that should have any effect. So we should try to improve the compiler instead.

File: base/broadcast.jl (hunk 1 above)
Comment (mateuszbaran) [reply base/broadcast.jl:372]: IIUC the compiler currently has problems with inlining recursive method calls, so by having two different methods doing the same thing there is less recursion to handle. There is no other reason to having two methods there. I don't know how exactly this works so I may be wrong though -- but it works.

**Pull request closed without merging**